### PR TITLE
perf: Combine small chunks in sinks for streaming pipelines

### LINKED ISSUE #11699: `sink_parquet(...)` much slower than `.collect().write_parquet(...)`
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

Using the same data as in https://github.com/pola-rs/polars/issues/11698, I'm observing at least a 2x performance difference in the following:

```python
pl.scan_parquet(f).sink_parquet("/tmp/out")

# 2x faster:
pl.scan_parquet(f).collect().write_parquet("/tmp/out")
```

Interestingly when looking at htop, the version with `.collect()` seems to use all cores sometimes (during reading?) while the other version never uses more than ~ 1 core.

This is a reduced example; the original tables has many more columns and the performance difference is much larger.

### Log output

_No response_

### Issue description

See example

### Expected behavior

Should have similar performance.

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.8
Index type:          UInt32
Platform:            macOS-13.5.2-arm64-arm-64bit
Python:              3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:41) [Clang 15.0.7 ]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
gevent:              <not installed>
matplotlib:          <not installed>
numpy:               1.25.0
openpyxl:            <not installed>
pandas:              2.0.3
pyarrow:             13.0.0
pydantic:            <not installed>
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by itamarst**
Fixes #11699 

Small chunks add significant overhead. This PR merges them in the sink. If there are lots of small chunks, this should make things faster. If there are lots of large chunks, it adds a little overhead but it's fixed and small per chunk, so that's fine. There may however be a slow down in some midway edge cases where the chunk merging involves copying extra data.

This doesn't fix the fact that small chunks are being generated, though... I will file a separate PR for the specific edge case in the reproducer script, and file an issue for another likely case if I can reproduce it.

### Benchmark

Before:

```
$ python ../11699.py 
SINK 2.8767316341400146
COLLECT+WRITE 0.9260973930358887
STREAMING COLLECT+WRITE 4.061856031417847
```

After:

```
$ python ../11699.py
SINK 1.064814805984497
COLLECT+WRITE 0.9025135040283203
STREAMING COLLECT+WRITE 1.1788344383239746
```

Using this script:

```python
from datetime import datetime

import polars as pl

start_datetime = datetime(2020, 1, 1)
end_datetime = datetime(
    2024,
    1,
    1,
)
N_ids = 10
interval = "1m"
df = (
    pl.LazyFrame(
        {
            "time": pl.datetime_range(
                start_datetime, end_datetime, interval=interval, eager=True
            )
        }
    )
    .join(
        pl.LazyFrame(
            {"id": [f"id{i:05d}" for i in range(N_ids)], "value": list(range(N_ids))}
        ),
        how="cross",
    )
    .select("time", "id", "value")
)

# print(df.profile(streaming=True))
from time import time

start = time()
df.sink_parquet("/tmp/out.parquet")
print("SINK", time() - start)

start = time()
df.collect().write_parquet("/tmp/out2.parquet")
print("COLLECT+WRITE", time() - start)

start = time()
df.collect(streaming=True).write_parquet("/tmp/out3.parquet")
print("STREAMING COLLECT+WRITE", time() - start)
```

### CODE REVIEW

File: crates/polars-pipe/src/operators/chunks.rs [hunk 1]
```diff
@@ -28,5 +28,152 @@ impl DataChunk {
...
+/// don't use too much memory by creating overly large allocations.
+///
+/// The assumption is that added `DataFrame`s are already in the correct order,
+/// and can therefore be combined.
+///
+/// 4 MB was chosen based on some empirical experiments that showed it to
+/// be decently faster than lower or higher values, and it's small enough
+/// it won't impact memory usage significantly.
+#[derive(Clone)]
+pub(crate) struct SemicontiguousVstacker<const BIG_ENOUGH: usize = 4194304> {
```
Comment (ritchie46) [inline crates/polars-pipe/src/operators/chunks.rs:62]: I don't think we have to make this a const parameter. That would require a recompile every instansiation that's not really worth it perf-wise.

File: crates/polars-pipe/src/operators/chunks.rs [hunk 2]
```diff
@@ -28,5 +28,152 @@ impl DataChunk {
...
+    current_dataframe: Option<DataFrame>,
+    /// Have we vstack()ed on to the current chunk?
+    stacked: bool,
+}
+
+impl<const BIG_ENOUGH: usize> SemicontiguousVstacker<BIG_ENOUGH> {
+    /// Create a new instance.
+    pub fn new() -> Self {
```
Comment (ritchie46) [inline crates/polars-pipe/src/operators/chunks.rs:70]: So then the `new` can accept the desired chunk size.

**Comment by itamarst** [comment]
Updates:

1. The failing test is plausibly not this PR's fault?
2. I tried to see if filtering could result in empirically measurable slowness. My initial attempt failed to show any effects, so I won't be filing a follow-up.

File: crates/polars-pipe/src/operators/chunks.rs [hunk 3]
```diff
@@ -28,5 +28,152 @@ impl DataChunk {
 }
 
 pub(crate) fn chunks_to_df_unchecked(chunks: Vec<DataChunk>) -> DataFrame {
-    accumulate_dataframes_vertical_unchecked(chunks.into_iter().map(|c| c.data))
+    let mut combiner = SemicontiguousVstacker::<4194304>::new();
```
Comment (ritchie46) [inline crates/polars-pipe/src/operators/chunks.rs:31]: I am not sure we must use that here. Let's first only fix the `writers`. 

For the operators, row-size is more important than size in megabytes.

**Review by ritchie46 — COMMENTED** [review]
Nice improvement. I have left some comments.

File: crates/polars-pipe/Cargo.toml [hunk 4]
```diff
@@ -31,6 +31,9 @@ smartstring = { workspace = true }
 [build-dependencies]
 version_check = { workspace = true }
 
+[dev-dependencies]
```
[the remark this answers is not in the document]
Comment (itamarst) [reply crates/polars-pipe/Cargo.toml:34]: I checked and this is already a requirement for some of the other crates. My guess is the Python tests would be a lot slower? And this class isn't really exposed...

File: crates/polars-pipe/src/operators/chunks.rs (hunk 1 above)
Comment (itamarst) [reply crates/polars-pipe/src/operators/chunks.rs:62]: It's just there for testing to save memory, but it's not necessary for testing, so I can remove.

File: crates/polars-pipe/src/operators/chunks.rs (hunk 3 above)
Comment (itamarst) [reply crates/polars-pipe/src/operators/chunks.rs:31]: I did it here in order to speed up `OrderedSink`, I can move it up the callstack so it's more isolated to only that.

File: crates/polars-pipe/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-pipe/Cargo.toml:34]: No, the python test run all in a single compilation. Every unit test in rust is compiled in a seperate `main` and costs a lot of time. Most of the time is in compilation time for our CI.

File: crates/polars-pipe/src/operators/chunks.rs (hunk 3 above)
Comment (ritchie46) [reply crates/polars-pipe/src/operators/chunks.rs:31]: Help me, why does this speed up `OrderdSink`? :thinking:

Comment (itamarst) [reply crates/polars-pipe/src/operators/chunks.rs:31]: Bad phrasing, sorry. It doesn't speed up OrderedSink, it speeds up writing the resulting DataFrame to parquet (the `collect(streaming=True).write_parquet()` case, `STREAMING COLLECT + WRITE` in the benchmark).

File: crates/polars-pipe/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (itamarst) [reply crates/polars-pipe/Cargo.toml:34]: Oh, but back to my original issue (thinking out loud):

1. This class isn't exposed to Python. It could be purely for testing purposes, I suppose... but that's making the public Python API larger with a private API. I guess could be prefixed with `_`.
2. Some sort of end-to-end test. This is... possible, I guess. vstack the DataFrames, turn into a `LazyFrame`, do `collect(streaming=False)` and then check that the chunks are smaller.

Do you have a preference?

[the remark this answers is not in the document]
Comment (itamarst) [reply crates/polars-pipe/Cargo.toml:34]: Also, are you sure about the fact that it's binary per `#[test]`? Reading https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html suggests that _integration tests_ do indeed get a binary per test, and discusses how to fix that, but _unit tests_ are a single binary per crate so the linking cost is one-time cost per crate.

[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-pipe/Cargo.toml:34]: Ok, but I am still not convinced we should do property tests on the rust side. Our CI isn't ready for this. We like to run fast unit-tests on our normal test runs and paramteric tests individually.

I can see use adding property tests on the rust side, but I want that set up then in CI properly first. Can you rewrite that test you wrote for only 3 inputs now and then in a seperate PR we can think of introducing property tests.

File: crates/polars-pipe/src/executors/sinks/ordered.rs [hunk 5]
```diff
@@ -53,8 +54,23 @@ impl Sink for OrderedSink {
             )));
         }
         self.sort();
+
         let chunks = std::mem::take(&mut self.chunks);
-        Ok(FinalizedSink::Finished(chunks_to_df_unchecked(chunks)))
+        let mut combiner = SemicontiguousVstacker::default();
```
Comment (ritchie46) [inline crates/polars-pipe/src/executors/sinks/ordered.rs:59]: I think a `rechunk_par` will be faster here. `vstack mut` will lead to reallocations on a quadratic strategy. Where a `rechunk` reallocates and knows how much to allocate up front. After that all can be memcopied.

I still think we shouldn't `rechunk` by default as now we always pay that cost even if you just want to see the results.

Most operations that come after will know if they need a rechunk or not to work optimal. So we shouldn't that decide for the consumer of this node.

File: crates/polars-pipe/src/executors/sinks/output/file_sink.rs [hunk 6]
```diff
@@ -40,13 +44,18 @@ pub(super) fn init_writer_thread(
                     chunks.sort_by_key(|chunk| chunk.chunk_index);
                 }
 
-                for chunk in chunks.iter() {
-                    writer._write_batch(&chunk.data).unwrap()
+                for chunk in chunks.drain(0..) {
```
Comment (ritchie46) [inline crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: Very crude solution. Can you test how this will perform with a single threaded `rechunk`?

Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: The issue is that if you just do a rechunk of `chunks`, you're limited to merging by the number of cores. E.g. with 12 cores, you merge 12 chunks, and that's all you merge. If you have 2 cores, you only merge 2 chunks, so the performance ends up being much worse if you have a stream of many small chunks. I can do another round of benchmarks tomorrow demonstrating that, but that's my memory from previous testing that led me to this strategy.

File: crates/polars-pipe/src/executors/sinks/ordered.rs (hunk 5 above)
Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/ordered.rs:59]: 1. I will try `rechunk_par` tomorrow.
2. I ... am not sure it's currently quadratic since there's only one rechunking per 4MB block, in `SemicontiguousVstacker::flush()`? But my mental model of `vstack_mut` may be wrong, but I thought it was just pointer copying until the point where you actually rechunk?
3. `write_parquet()` is definitely _not_ rechunking, thus the problem. So another option is moving the rechunking there instead of here.

File: crates/polars-pipe/src/executors/sinks/output/file_sink.rs (hunk 6 above)
Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: Notice that `chunks` size is `morsels_per_sink` which is the size of the thread pool, basically. So the fewer cores you have, the less merging only `chunks` helps.

File: crates/polars-pipe/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (itamarst) [reply crates/polars-pipe/Cargo.toml:34]: Will do.

File: crates/polars-pipe/src/executors/sinks/ordered.rs (hunk 5 above)
Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/ordered.rs:59]: 1. I assume you mean `as_single_chunk_par()`?
2. I remembered why I'm not doing this: the goal is _not_ to do a single chunk, cause that means a huge allocation, doubling memory usage. Whereas doing e.g. 4MB at a time means you effectively have the same memory usage, while still getting the benefit of having none-tiny chunks.

To rephrase: this PR is a half-way point between "stick to a massive number of tiny chunks" and "rechunk into a single chunk", where it tries to have no real significant increase in memory requirements, and hopefully not too much of a performance hit.

File: crates/polars-pipe/src/executors/sinks/output/file_sink.rs (hunk 6 above)
Comment (ritchie46) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: > f you have 2 cores, you only merge 2 chunks, so the performance ends up being much worse if you have a stream of many small chunks. I can do another round of benchmarks tomorrow demonstrating that, but that's my memory from previous testing that led me to this strategy.

I don't really follow, because a `vstack_mut` is essentially the same. It only has one downside that is that we cannot know the size to allocate. So we can trigger multiple reallocations. With a `rechunk` we first do a pass to determine the required output memory size and then we trigger the memcopies. That's the reason I think we should just `rechunk`.

Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: So I took this as an opportunity to learn Divan, it's pretty cool: https://github.com/itamarst/benchmarks/blob/main/polars-vstack/benches/vstack.rs#L13

Running this I get:

```
Timer precision: 19 ns
vstack        fastest       │ slowest       │ median        │ mean          │ samples │ iters
╰─ vstack                   │               │               │               │         │
   ├─ 10000   885.2 µs      │ 1.049 ms      │ 921.5 µs      │ 925.5 µs      │ 100     │ 100
   │          90.36 Gitem/s │ 76.22 Gitem/s │ 86.8 Gitem/s  │ 86.43 Gitem/s │         │
   │          alloc:        │               │               │               │         │
   │            10009       │ 10009         │ 10009         │ 10009         │         │
   │            1.216 MB    │ 1.216 MB      │ 1.216 MB      │ 1.216 MB      │         │
   │          dealloc:      │               │               │               │         │
   │            2           │ 2             │ 2             │ 2             │         │
   │            8.084 KB    │ 8.084 KB      │ 8.084 KB      │ 8.084 KB      │         │
   │          grow:         │               │               │               │         │
   │            13          │ 13            │ 13            │ 13            │         │
   │            262.1 KB    │ 262.1 KB      │ 262.1 KB      │ 262.1 KB      │         │
   ╰─ 100000  9.565 ms      │ 9.966 ms      │ 9.696 ms      │ 9.713 ms      │ 100     │ 100
              83.63 Gitem/s │ 80.26 Gitem/s │ 82.5 Gitem/s  │ 82.35 Gitem/s │         │
              alloc:        │               │               │               │         │
                100009      │ 100009        │ 100009        │ 100009        │         │
                12.01 MB    │ 12.01 MB      │ 12.01 MB      │ 12.01 MB      │         │
              dealloc:      │               │               │               │         │
                2           │ 2             │ 2             │ 2             │         │
                8.084 KB    │ 8.084 KB      │ 8.084 KB      │ 8.084 KB      │         │
              grow:         │               │               │               │         │
                16          │ 16            │ 16            │ 16            │         │
                2.097 MB    │ 2.097 MB      │ 2.097 MB      │ 2.097 MB      │         │
```

I'm really not getting the impression that repeated `vstack_mut` leads to quadratic slowdown. Runtime is linear, allocated memory is linear and much smaller than underlying data: 100_000 * 1000 * 8 bytes is 800MB, but only 12MB gets allocated.

Next going to read the code for `vstack_mut` so I understand implementation better.

Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: So AFAICT `vstack_mut` eventually devolves to `new_chunks()` in `chunked_array/ops/append.rs`, which clones all the chunks individually, i.e. does not make them contiguous.

Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: I think we might be talking past each other here? So let me try again, and you can point out what if anything I got wrong.

Basically this PR is structured the way it is because a separate struct is necessary to accumulate state across batches.

The way `init_writer_thread` currently works:

1. It reads from a queue and adds `DataChunk` to `chunks`, a vec.
2. After receiving N `DataChunk`, where N is basically "the number of threads in the thread pool" it optionally sorts them, then writes the batch one by one to the underlying writer.

We know the individual chunks are too small, and we don't want to write them one by one when they're small, cause that's slow. So we want to combine them, somehow. There are two strategies one could take:

1. **Combine the batch:** Combine everything in `chunks` into a single DataFrame. This means we can combine at most N chunks at a time, no more, and for smaller thread pool sizes this might only be 2 chunks!
2. **Combine across multiple batches if necessary.** This is what this PR does. When you call `SemicontiguousVstacker.add()`, it might return nothing and keep the data in its internal buffer! So it can continue accumulating across write batches until it hits some limit (in current state, 4MB of data). For a 2.5KB-sized chunk, that means it's accumulating about 1600 chunks. That's a lot more than 12 at a time!

**My understanding is that you are proposing strategy 1, combine the single batch.** If that understanding is wrong, I'd appreciate a fuller explanation or code sketch of what you mean.

### Benchmarking the first strategy

I used the following:

```
            if chunks.len() == morsels_per_sink || last_write {
                if maintain_order {
                    chunks.sort_by_key(|chunk| chunk.chunk_index);
                }

                 if !chunks.is_empty() {
                    let mut merged = chunks_to_df_unchecked(std::mem::take(&mut chunks));
                    merged.as_single_chunk();
                    writer._write_batch(&merged).unwrap();
                }

                if last_write {
                    writer._finish().unwrap();
                    return;
                }
            }
```

Now, if run this:

```
$ python ../11699.py 
SINK 1.1473236083984375
COLLECT+WRITE 0.9269287586212158
```

Pretty good! However, if the thread pool is smaller, the code runs twice as slow, because of parquet writing of small chunks:

```
$ POLARS_MAX_THREADS=2 python ../11699.py 
SINK 2.1068201065063477
COLLECT+WRITE 0.9157388210296631
```

In contrast, this PR is unaffected by thread pool size because the rechunking is done by data size, not by number of morsels (== length of `chunks` == thread pool size).

Comment (ritchie46) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: > I'm really not getting the impression that repeated vstack_mut leads to quadratic slowdown

No, I don't mean quadratic slowdown. It is a quadratic growing strategy. Which means we might have a few reallocs.

Alright. What I mean is that we need a combination of the strategies. So:

1. Batch chunks across multiple batches. Agree on that
2. Combine them into a single chunk. Also agree on that, but I think we should use `rechunk` (all batches just before flush) instead of `vstack_mut` here.

Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: Maybe I'm missing something in the API search, but `rechunk()` seems to be specific to a single `Series`/`DataFrame`? Which would help if the series of `DataChunk` contents was:

```
Series<[2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk]>
Series<[2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk]>
Series<[2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk]>
...
```

But actually the content is:

```
Series<[2.5K chunk]>
Series<[2.5K chunk]>
Series<[2.5K chunk]>
Series<[2.5K chunk]>
...
```

So `rechunk()` on an individual `Series` won't help, you need to combine them somehow, and `vstack_mut` (or perhaps the lower cost `vstack_mut_unchecked`, if it were public) seems to be the only option.

Comment (ritchie46) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: Yes. you first combine them into a single `DataFrame` with multiple chunks and then you `rechunk`. Depening on the sink this should be done parallel (`as_single_chunk_par`) or single threaded `as_single_chunk`. Here we should do it single threaded.

So I agree with most of the code here, I just want to use `rechunk` as it allocates properly. If you want I can finish the PR.
So

Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/output/file_sink.rs:47]: Re-reading all this again, my impression is that you want the _caller_ to decide whether and how to rechunk, instead of, as is currently implemented, `SemicontiguousVstacker` having a fixed policy of calling `as_single_chunk()`.

I tested the code with `as_single_chunk_par`. It appeared to cause a deadlock in the `sink_parquet()` case :grimacing: which I didn't investigate further. It did indeed make the `collect(streaming=True).write_parquet()` version slightly faster.

So I will try implement the above, and if it's still not you meant please do finish the PR, yes.

**Comment by itamarst** [comment]
@ritchie46 OK hopefully I've finally understood you, sorry it took so long. If I didn't, please feel free to finish this PR.

Final numbers:

```
SINK 1.0646779537200928
COLLECT+WRITE 0.9343609809875488
STREAMING COLLECT+WRITE 1.083845853805542
```

I.e. the `collect(streaming=True).write_parquet()` case is slightly faster due to the `as_single_chunk_par()` change.

File: crates/polars-pipe/src/executors/sinks/ordered.rs [hunk 7]
```diff
@@ -57,32 +55,7 @@ impl Sink for OrderedSink {
         self.sort();
 
         let chunks = std::mem::take(&mut self.chunks);
-        let mut combiner = StreamingVstacker::default();
```
Comment (ritchie46) [inline crates/polars-pipe/src/executors/sinks/ordered.rs]: If we would rechunk, we could simply rechunk here. But I don't want to do that as that should be left to the consumer of the streaming engine.

Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/ordered.rs]: In that case it should probably be done in `write_parquet()`, otherwise the `collect(streaming=True).write_parquet()` case will continue to be slow.

Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/ordered.rs]: Which would requite the new struct be e.g. moved into the `polars-core` crate and made public.

Here's the runtime with latest commit on my computer, last case is slow again:

```
SINK 1.0721302032470703
COLLECT+WRITE 0.9235994815826416
STREAMING COLLECT+WRITE 4.254129648208618
```

Comment (itamarst) [reply crates/polars-pipe/src/executors/sinks/ordered.rs]: But maybe also `write_feather()`. or the cloud parquet writer. etc.. (Having it in OrderedSink seemed like a low-cost smoothing of performance bumps, limited to a single place.)

Comment (ritchie46) [reply crates/polars-pipe/src/executors/sinks/ordered.rs]: Then that logic should be in `write_parquet` indeed. That writer should check the chunk sizes.

I will first merge this one, and then we can follow up with the `write_parquet` optimization.

Comment (ritchie46) [reply crates/polars-pipe/src/executors/sinks/ordered.rs]: > But maybe also `write_feather()`. or the cloud parquet writer. etc.. (Having it in OrderedSink seemed like a low-cost smoothing of performance bumps, limited to a single place.)

Yes, but it is more expensive for other operations. Operations themselves should knie their best chunking strategies.

**Comment by itamarst** [comment]
Thanks for all the feedback and help, and yes, I'm happy to do follow-up PR for `write_parquet()`.
